stdlib/set.ml

```diff
@@ -609,15 +625,169 @@ module Make(Ord: OrderedType) =
-              | 0 -> More (v, r, c)
-              | n when n<0 -> aux low r c
-              | _ -> aux low l (More (v, r, c))
-            end
-      in
-      seq_of_enum_ (aux low s End)
+      seq_of_enum_ (enum_from low s)
+
+    module Enum = struct
+
+      (* The type [enumeration] is the same as above: *)
+
+      (* type enumeration = End | More of elt * t * enumeration *)
+
+      (* An enumeration satisfies the following invariant: it is a path (that
+         is, a zipper) inside a balanced binary search tree [t]. It is a path
+         whose edges point up, that is, a path from some internal node up to
+         the root of the tree [t]. Along this path, every edge points towards
+         the right, that is, from a left child to its parent. The enumeration
+         [More (v, r, e)] can be understood as an edge from a left child to a
+         parent that carries the value [v], the subtree [r], and whose own
+         parent is [e]. The path edges that point towards the left (that is,
+         from a right child to its parent) are not materialized; they do not
+         appear in an enumeration.
+
+         This invariant implies that the length of an enumeration, viewed as a
+         list of pairs (x_i, t_i) of an element and a subtree, is logarithmic:
+         it is O(log n) where [n] is the size of the tree [t].
+
+         Furthermore, this invariant implies that the subtrees t_1, t_2, ...,
+         t_k exist at depths d_1 > d_2 > ... > d_k in the tree [t]. Because
+         the tree [t] is balanced, this implies that the sizes of the subtrees
+         t_1, t_2, ..., t_k obey a (rough) geometric progression: that is, the
+         size of the subtree t_{i+1} is (roughly) at least twice the size of
+         the subtree t_i.
+
+         The code that follows does not care exactly how trees are balanced,
+         as long as they are balanced in some way. (For example, they could be
+         height-balanced or weight-balanced.) Except for the function
+         [elements], this code does not construct any tree; instead, it
+         offers facilities for iterating over an existing tree.
+
+         The code that follows does not directly construct any enumerations.
+         It constructs enumerations indirectly by invoking the functions
+         [cons_enum], [enum_from_aux], and [enum_from], which are defined
+         higher up in this file.
+
+         The time complexity of [from low e] should be O(log d) where [d] is
+         the distance traveled, that is, the number of elements that are
+         dropped from the enumeration. *)
+
+      type tree = t
+
+      type set = tree
+
+      type enum = enumeration
+
+      type t = enum
+
+      let empty : enum =
+        End
+
+      let is_empty (e : enum) : bool =
+        match e with End -> true | More _ -> false
+
+      let[@inline] enum (t : tree) : enum =
+        cons_enum t End
+          (* [cons_enum] was defined earlier in this file *)
```

Comment:
I didn't look at the implementation in details but I noticed quite a few of these comments in the file. They look useless to me. I think we can assume that people reading this code are aware of OCaml's scoping rules :-)